File: lib/utils/content_hash.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Must match `backend/src/utils/contentHash.js` (citizen).
String hashCitizenQr({
  required String nationalId,
  required String serviceId,
  required String payload,
  required String createdAt,
}) {
  final s = '$nationalId|$serviceId|$payload|$createdAt';
  final d = sha256.convert(utf8.encode(s));
  return d.toString();
}

/// Must match `backend/src/utils/contentHash.js` (agent app uses same for scans).
String hashAgentScan({
  String? userId,
  required String rawPayload,
  required String scannedAt,
}) {
  final s = 'agent|${userId ?? ''}|$rawPayload|$scannedAt';
  final d = sha256.convert(utf8.encode(s));
  return d.toString();
}

File: backend/src/utils/contentHash.js
import crypto from 'node:crypto';

/**
 * Must match `lib/utils/content_hash.dart` (citizen).
 */
export function hashCitizenQr({ nationalId, serviceId, payload, createdAt }) {
  const s = `${nationalId}|${serviceId}|${payload}|${createdAt}`;
  return crypto.createHash('sha256').update(s, 'utf8').digest('hex');
}

/**
 * Must match `lib/utils/content_hash.dart` (agent).
 */
export function hashAgentScan({ userId, rawPayload, scannedAt }) {
  const s = `agent|${userId ?? ''}|${rawPayload}|${scannedAt}`;
  return crypto.createHash('sha256').update(s, 'utf8').digest('hex');
}
